package org.cbioportal.models;

public class Header {
    public static final String Hugo_Symbol = "Hugo_Symbol";
    public static final String Entrez_Gene_Id = "Entrez_Gene_Id";
    public static final String Center = "Center";
    public static final String NCBI_Build = "NCBI_Build";
    public static final String Chromosome = "Chromosome";
    public static final String Start_Position = "Start_Position";
    public static final String End_Position = "End_Position";
    public static final String Strand = "Strand";
    public static final String Consequence = "Consequence";
    public static final String Variant_Classification = "Variant_Classification";
    public static final String Variant_Type = "Variant_Type";
    public static final String Reference_Allele = "Reference_Allele";
    public static final String Tumor_Seq_Allele1 = "Tumor_Seq_Allele1";
    public static final String Tumor_Seq_Allele2 = "Tumor_Seq_Allele2";
    public static final String dbSNP_RS = "dbSNP_RS";
    public static final String dbSNP_Val_Status = "dbSNP_Val_Status";
    public static final String Tumor_Sample_Barcode = "Tumor_Sample_Barcode";
    public static final String Matched_Norm_Sample_Barcode = "Matched_Norm_Sample_Barcode";
    public static final String Match_Norm_Seq_Allele1 = "Match_Norm_Seq_Allele1";
    public static final String Match_Norm_Seq_Allele2 = "Match_Norm_Seq_Allele2";
    public static final String Tumor_Validation_Allele1 = "Tumor_Validation_Allele1";
    public static final String Tumor_Validation_Allele2 = "Tumor_Validation_Allele2";
    public static final String Match_Norm_Validation_Allele1 = "Match_Norm_Validation_Allele1";
    public static final String Match_Norm_Validation_Allele2 = "Match_Norm_Validation_Allele2";
    public static final String Verification_Status = "Verification_Status";
    public static final String Validation_Status = "Validation_Status";
    public static final String Mutation_Status = "Mutation_Status";
    public static final String Sequencing_Phase = "Sequencing_Phase";
    public static final String Sequence_Source = "Sequence_Source";
    public static final String Validation_Method = "Validation_Method";
    public static final String Score = "Score";
    public static final String BAM_File = "BAM_File";
    public static final String Sequencer = "Sequencer";
    public static final String t_ref_count = "t_ref_count";
    public static final String t_alt_count = "t_alt_count";
    public static final String n_ref_count = "n_ref_count";
    public static final String n_alt_count = "n_alt_count";
    public static final String HGVSc = "HGVSc";
    public static final String HGVSp = "HGVSp";
    public static final String HGVSp_Short = "HGVSp_Short";
    public static final String Transcript_ID = "Transcript_ID";
    public static final String RefSeq = "RefSeq";
    public static final String Protein_position = "Protein_position";
    public static final String Codons = "Codons";
    public static final String Genomic_Change = "Genomic_Change";
    public static final String Probabilistic_Allelic_Frequency = "Probabilistic_Allelic_Frequency";
    public static final String Secondary_Variant_Classification = "Secondary_Variant_Classification";
    public static final String Transcript = "Transcript";
    public static final String Transcript_Exon = "Transcript_Exon";
    public static final String Transcript_Position = "Transcript_Position";
    public static final String Transcript_Strand = "Transcript_Strand";
    public static final String cdna_change = "cdna_change";
    public static final String cds_change = "cds_change";
    public static final String gc_content = "gc_content";
    public static final String protein_change = "protein_change";
    public static final String reference_context = "reference_context";
    public static final String ssm2_pass_call = "ssm2_pass_call";
    public static final String Annotation_Status = "Annotation_Status";
    public static final String Exon_Number = "Exon_Number";

}
